### PR TITLE
Add axis for boxcoder

### PR DESCRIPTION
**Pull request by jerrywgz**

### CODE REVIEW

File: paddle/fluid/operators/detection/box_coder_op.cc [hunk 1]
```diff
@@ -32,32 +32,57 @@ class BoxCoderOp : public framework::OperatorWithKernel {
 
     if (ctx->IsRuntime()) {
       PADDLE_ENFORCE_EQ(prior_box_dims.size(), 2,
-                        "The rank of Input of PriorBoxVar must be 2");
+                        "The rank of Input of PriorBox must be 2");
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/box_coder_op.cc]: The rank of Input of PriorBox ->  The rank of Input(PriorBox)

File: paddle/fluid/operators/detection/box_coder_op.cc [hunk 2]
```diff
@@ -32,32 +32,57 @@ class BoxCoderOp : public framework::OperatorWithKernel {
...
-        PADDLE_ENFORCE_EQ(target_box_dims[1], 4,
-                          "The shape of TargetBox is [M, 4]");
-      } else if (code_type == BoxCodeType::kDecodeCenterSize) {
-        PADDLE_ENFORCE_EQ(target_box_dims.size(), 3,
-                          "The rank of Input of TargetBox must be 3");
+    auto code_type = GetBoxCodeType(ctx->Attrs().Get<std::string>("code_type"));
+    int axis = ctx->Attrs().Get<int>("axis");
+    if (code_type == BoxCodeType::kEncodeCenterSize) {
+      PADDLE_ENFORCE_EQ(target_box_dims.size(), 2,
+                        "The rank of Input of TargetBox must be 2");
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/box_coder_op.cc]: The rank of Input of TargetBox ->  The rank of Input(TargetBox)

File: paddle/fluid/operators/detection/box_coder_op.cc [hunk 3]
```diff
@@ -32,32 +32,57 @@ class BoxCoderOp : public framework::OperatorWithKernel {
...
+      PADDLE_ENFORCE_EQ(target_box_dims[1], 4,
+                        "The shape of TargetBox is [M, 4]");
+      ctx->SetOutputDim(
+          "OutputBox",
+          framework::make_ddim({target_box_dims[0], prior_box_dims[0], 4}));
+    } else if (code_type == BoxCodeType::kDecodeCenterSize) {
+      PADDLE_ENFORCE_EQ(target_box_dims.size(), 3,
+                        "The rank of Input of TargetBox must be 3");
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/box_coder_op.cc]: The rank of Input of TargetBox  -> The rank of Input(TargetBox)

File: paddle/fluid/operators/detection/box_coder_op.cc [hunk 4]
```diff
@@ -100,6 +125,12 @@ class BoxCoderOpMaker : public framework::OpProtoAndCheckerMaker {
                   "(bool, default true) "
                   "whether treat the priorbox as a noramlized box")
         .SetDefault(true);
+    AddAttr<int>("axis",
+                 "(int, default 1)"
+                 "which axis to broadcast for box decode, it is only valid"
+                 "when code type is decode_center_size")
+        .SetDefault(0)
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/box_coder_op.cc:136]: line 129和line132行的default不一致

File: paddle/fluid/operators/detection/box_coder_op.cc [hunk 5]
```diff
@@ -100,6 +125,12 @@ class BoxCoderOpMaker : public framework::OpProtoAndCheckerMaker {
                   "(bool, default true) "
                   "whether treat the priorbox as a noramlized box")
         .SetDefault(true);
+    AddAttr<int>("axis",
+                 "(int, default 1)"
+                 "which axis to broadcast for box decode, it is only valid"
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/box_coder_op.cc]: > which axis to broadcast for box decode

解释的不明白，对谁broadcast，怎么broadcast

File: paddle/fluid/operators/detection/box_coder_op.cu [hunk 6]
```diff
@@ -30,11 +31,9 @@ __global__ void EncodeCenterSizeKernel(const T* prior_box_data,
...
                          prior_box_data[col_idx * len + 1] +
                          (normalized == false);
-    T prior_box_center_x =
-        (prior_box_data[col_idx * len + 2] + prior_box_data[col_idx * len]) / 2;
-    T prior_box_center_y = (prior_box_data[col_idx * len + 3] +
-                            prior_box_data[col_idx * len + 1]) /
-                           2;
+    T prior_box_center_x = prior_box_data[col_idx * len] + prior_box_width / 2;
+    T prior_box_center_y =
+        prior_box_data[col_idx * len + 1] + prior_box_height / 2;
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/box_coder_op.cu:38]: Why change this computing?

File: paddle/fluid/operators/detection/box_coder_op.cu [hunk 7]
```diff
@@ -68,33 +71,48 @@ __global__ void DecodeCenterSizeKernel(const T* prior_box_data,
...
     const int col_idx = idx % col;
-    T prior_box_width = prior_box_data[col_idx * len + 2] -
-                        prior_box_data[col_idx * len] + (normalized == false);
-    T prior_box_height = prior_box_data[col_idx * len + 3] -
-                         prior_box_data[col_idx * len + 1] +
+    const int row_idx = idx / col;
+    if (axis == 0)
+      prior_box_offset = col_idx * len;
+    else if (axis == 1)
+      prior_box_offset = row_idx * len;
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/box_coder_op.cu]: use: `? :`
```c++
prior_box_offset =  axis == 0 ?  : ;
```

File: paddle/fluid/operators/detection/box_coder_op.cu [hunk 8]
```diff
@@ -68,33 +71,48 @@ __global__ void DecodeCenterSizeKernel(const T* prior_box_data,
...
     T target_box_width, target_box_height;
     T target_box_center_x, target_box_center_y;
     if (prior_box_var_data) {
-      target_box_width = exp(prior_box_var_data[col_idx * len + 2] *
+      int prior_var_offset = 0;
+      if (prior_box_var_size == 2) {
+        if (axis == 0)
+          prior_var_offset = col_idx * len;
+        else if (axis == 1)
+          prior_var_offset = row_idx * len;
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/box_coder_op.cu]: same as above

File: python/paddle/fluid/layers/detection.py [hunk 9]
```diff
@@ -342,6 +342,7 @@ def box_coder(prior_box,
               target_box,
               code_type="encode_center_size",
               box_normalized=True,
+              axis=0,
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: put it in last line.

File: python/paddle/fluid/tests/unittests/test_box_coder_op.py [hunk 10]
```diff
@@ -49,26 +59,52 @@ def box_coder(target_box, prior_box, prior_box_var, output_box, code_type,
...
-        output_box[:,:,1] = (target_box_y - prior_box_y) / prior_box_height / \
-                prior_box_var[:,:,1]
-        output_box[:,:,2] = np.log(np.fabs(target_box_width / prior_box_width)) / \
-                prior_box_var[:,:,2]
-        output_box[:,:,3] = np.log(np.fabs(target_box_height / prior_box_height)) / \
-                prior_box_var[:,:,3]
+        if prior_box_var.ndim == 1:
+            output_box[:,:,0] = (target_box_x - prior_box_x) / \
+                                prior_box_width / \
+                                prior_box_var[0]
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_box_coder_op.py]: 1. 最好简化逻辑， encode, decoder分成不同的函数。一个函数代码行数短。
2. 这里变量名字其实可以简化，并能读懂，尽量使得一行可以放下，而不是用 \

比如 
target_box_x -> t_x
prior_box_x -> p_x
prior_box_width -> p_w

File: paddle/fluid/operators/detection/box_coder_op.cc [hunk 11]
```diff
@@ -32,32 +32,57 @@ class BoxCoderOp : public framework::OperatorWithKernel {
...
         auto prior_box_var_dims = ctx->GetInputDim("PriorBoxVar");
-        PADDLE_ENFORCE_EQ(prior_box_dims, prior_box_var_dims);
+        PADDLE_ENFORCE(
+            prior_box_var_dims.size() == 1 || prior_box_var_dims.size() == 2,
+            "Input(PriorBoxVar) of BoxCoderOp should be 1 or 2.");
+        if (prior_box_var_dims.size() == 1) {
+          PADDLE_ENFORCE_EQ(
+              prior_box_var_dims[0], 4,
+              "The 1st dimension of Input(PriorBoxVar) should be 4"
+              "when the rank is 1.");
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/box_coder_op.cc:48]: 如果只有4个数， 可以用attr吧？

File: python/paddle/fluid/tests/unittests/test_box_coder_op.py [hunk 12]
```diff
@@ -21,22 +21,32 @@
...
-
+def box_coder(target_box,
+              prior_box,
+              prior_box_var,
+              output_box,
+              code_type,
+              box_normalized,
+              axis=0):
+    prior_box_width = prior_box[:, 2] - prior_box[:, 0] + \
+                      (box_normalized==False)
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_box_coder_op.py]: 同上，尽量少用 \换行， 比如可以 box_normalized ->  norm， 命名短一些

File: paddle/fluid/operators/detection/box_coder_op.cc (hunk 1 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/box_coder_op.cc]: Done

File: paddle/fluid/operators/detection/box_coder_op.cc (hunk 2 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/box_coder_op.cc]: Done

File: paddle/fluid/operators/detection/box_coder_op.cc (hunk 3 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/box_coder_op.cc]: Done

File: paddle/fluid/operators/detection/box_coder_op.cc (hunk 4 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/box_coder_op.cc]: Done

File: paddle/fluid/operators/detection/box_coder_op.cc (hunk 5 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/box_coder_op.cc]: Done

Comment (jerrywgz) [reply paddle/fluid/operators/detection/box_coder_op.cc]: Done

File: paddle/fluid/operators/detection/box_coder_op.cc (hunk 4 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/box_coder_op.cc:136]: Done

File: paddle/fluid/operators/detection/box_coder_op.cu (hunk 6 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/box_coder_op.cu:38]: When boxes are normalized, the formula is same, while when boxes are not normalized, the old formula has diff compared to Detectron

File: paddle/fluid/operators/detection/box_coder_op.cu (hunk 7 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/box_coder_op.cu]: Done

File: paddle/fluid/operators/detection/box_coder_op.cu (hunk 8 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/box_coder_op.cu]: Done

File: python/paddle/fluid/tests/unittests/test_box_coder_op.py (hunk 12 above)
Comment (jerrywgz) [reply python/paddle/fluid/tests/unittests/test_box_coder_op.py]: Done

File: python/paddle/fluid/layers/detection.py (hunk 9 above)
Comment (jerrywgz) [reply python/paddle/fluid/layers/detection.py]: Done

File: python/paddle/fluid/tests/unittests/test_box_coder_op.py (hunk 10 above)
Comment (jerrywgz) [reply python/paddle/fluid/tests/unittests/test_box_coder_op.py]: Done

File: paddle/fluid/operators/detection/box_coder_op.cc (hunk 11 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/box_coder_op.cc:48]: 这里box var可以为4个数，也接受dim size不为1的情况

File: python/paddle/fluid/tests/unittests/test_box_coder_op.py [hunk 13]
```diff
@@ -111,6 +111,32 @@ def setUp(self):
...
+class TestBoxCoderOpWithOneRankVar(OpTest):
+    def test_check_output(self):
+        self.check_output()
+
+    def setUp(self):
+        self.op_type = "box_coder"
+        lod = [[1, 1, 1, 1, 1]]
+        prior_box = np.random.random((6, 4)).astype('float32')
+        prior_box_var = np.random.random((4)).astype('float32')
+        target_box = np.random.random((3, 6, 4)).astype('float32')
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_box_coder_op.py]: Better use large shape to test the boundary condition of  CUDA kernel

same as follows

**Review by qingqing01 — COMMENTED** [review]
LGTM

File: paddle/fluid/operators/detection/box_coder_op.cu [hunk 14]
```diff
@@ -127,18 +156,39 @@ class BoxCoderCUDAKernel : public framework::OpKernel<T> {
...
+                     "Size of attribute 'variance' should be 4");
+    }
 
     if (target_box->lod().size()) {
       PADDLE_ENFORCE_EQ(target_box->lod().size(), 1,
                         "Only support 1 level of LoD.");
     }
+    const int var_size = static_cast<T>(variance.size());
+    thrust::device_vector<float> dev_variance(variance.begin(), variance.end());
+    const float* dev_var_data = thrust::raw_pointer_cast(dev_variance.data());
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/box_coder_op.cu]: Do not use `thrust::device_vector`. Please use TempryAllocator, you can refer roi_align/pool. 

Or since only 4 values, passed in the input arguments.

File: paddle/fluid/operators/detection/box_coder_op.h [hunk 15]
```diff
@@ -78,57 +79,94 @@ class BoxCoderKernel : public framework::OpKernel<T> {
...
+                                prior_box_center_y;
+          target_box_width = std::exp(prior_box_var_data[prior_var_offset + 2] *
+                                      target_box_data[offset + 2]) *
+                             prior_box_width;
+          target_box_height =
+              std::exp(prior_box_var_data[prior_var_offset + 3] *
+                       target_box_data[offset + 3]) *
+              prior_box_height;
+        } else if (!(variance.empty())) {
+          target_box_center_x = static_cast<T>(variance[0]) *
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/box_coder_op.h]: 避免一个函数的代码太长，可以 var.size==4，写成外一个函数。

File: paddle/fluid/operators/detection/box_coder_op.cu [hunk 16]
```diff
@@ -55,46 +55,75 @@ __global__ void EncodeCenterSizeKernel(const T* prior_box_data,
...
+      target_box_width =
+          exp(static_cast<T>(variance[2]) * target_box_data[idx * len + 2]) *
+          prior_box_width;
+      target_box_height =
+          exp(static_cast<T>(variance[3]) * target_box_data[idx * len + 3]) *
+          prior_box_height;
+      target_box_center_x = static_cast<T>(variance[0]) *
+                                target_box_data[idx * len] * prior_box_width +
+                            prior_box_center_x;
+      target_box_center_y = static_cast<T>(variance[1]) *
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/box_coder_op.cu]: 同上，==4的分支，可以写成另外一个函数。

File: python/paddle/fluid/tests/test_detection.py [hunk 17]
```diff
@@ -59,7 +59,7 @@ def test_detection_api(self):
             iou = layers.iou_similarity(x=x, y=y)
             bcoder = layers.box_coder(
                 prior_box=x,
-                prior_box_var=y,
+                prior_box_var=[0.2, 0.3, 0.3, 0.2],
```
Comment (qingqing01) [inline python/paddle/fluid/tests/test_detection.py]: 可以新加一个测试，而不是改原来的。

Comment (jerrywgz) [reply python/paddle/fluid/tests/test_detection.py]: Done

File: paddle/fluid/operators/detection/box_coder_op.h (hunk 15 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/box_coder_op.h]: 简化了三个分支

File: paddle/fluid/operators/detection/box_coder_op.cu (hunk 14 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/box_coder_op.cu]: Done

File: paddle/fluid/operators/detection/box_coder_op.cu (hunk 16 above)
Comment (jerrywgz) [reply paddle/fluid/operators/detection/box_coder_op.cu]: 同样简化了三个分支

**Review by qingqing01 — APPROVED** [review]

**Review by panyx0718 — APPROVED** [review]
